### ISSUE TITLE
Previews

### ISSUE THREAD

**Issue by iaincollins** [alephdata/aleph#340]
On Friday we came up with an approach we'd like to follow for previews in the search view.

This is an attempt to capture the specifics of the behaviour we want and to detail some of the tedious implementation details of the CSS/markup, to explain what we explored for posterity.

This is a bit stream of consciousness and I might tidy it up and trim it down later to be more concise, I just wanted to capture the out come of discussions on Friday (and over the weekend) around it.

* When you select an entity or document a sidebar should appear.
* The sidebar should slide in from the right.
* It should overlay content on the search results, not reflow it.
* It should have a close button that hides the Preview.
* It should have an initial fixed with (~400px) but also a maximise button.
* The maximise button should toggle between the small fixed with and ~50% of the window size (possibly with a max-width constraint of ~1024px).
* When in a non-maximised state it should display summary information.
* When maximised it should be able to display the actual document or entity being previewed.
* It should be possible to use arrow keys / buttons to navigate to the next / previous document in a search result.
* The item being previewed in the search result should be highlighted while the preview is open so it's clear to the user what happens when they select to preview the next document.
* The component should be included at the Screen level (i.e. in Screen.jsx).
* It should use URL reading and writing (of the # path) to determine if it is visible, what content it should display, what size it should display at if open and what tab is selected.
* The # path schema is TBD but, for example: #preview:id=123&preview:type=entity&preview:size=maximised&preview:tab=overview

# CSS Implementation

I have built three prototypes now and think I have some idea about our options for CSS implementation.

## Sticky Sidebar

The CSS layout option `position: sticky` could suit our needs, but has some limitations. Some of those are being worked on in the CSS spec (such as a selector for knowing when an stick element has reached the top of the screen and become 'sticky' so internal content can be reflowed).

Most commonly, Javascript watchers are used to assist when rendering stick elements, including with libraries like Bootstrap which uses it for it's Navbar.

Examples:

* https://www.npmjs.com/package/react-sticky-el
* https://www.npmjs.com/package/react-sticky
* https://css-tricks.com/scroll-fix-content/ <- Older, but has good comments.

Perhaps the best module for React + Sticky is one written by Yahoo!. It is much less popular than some of the others, but I think better engineered. It's README goes into the short falls of some of the existing implementations and the issues and the edge cases it addresses. These are relevant to us.

The other libraries are great for most use cases, which are much simpler than ours (in particular, which don't involve things like Flex boxes or auto-expanding results pages) but we would need to use something like the Yahoo! version if we went down this route.

* https://github.com/yahoo/react-stickynode

## Fixed Sidebar

My experiments with different sticky approaches led me to `position: fixed`.

Essentially, I think it's the lesser of evils.

There are a several odd issues that pop up with position sticky, in particular with relation to using a flexbox layout. I think we could address these but it would complicate things significantly for maintenance and seems likely to be more fragile.

The Yahoo! library does a great job of abstracting the watchers and markup, but it still has to wrap all your components and does a lot of work behind the scenes.

Based on creating three separate prototypes - and our requirements above - I lean towards Fixed positioning - with a small amount of JavaScript - for the Preview pane. 

* When displaying initially it would need to account for  the px height of the Navbar and Footer. This can be done in real time when the component is displayed, so will display correctly even if we change the height of the Navbar or Footer. 

* It would need a debounced watcher to check scroll position so it behaves like a sticky element when the Navbar scrolls off the screen (so it takes up the full height of the window and doesn't leave a gap at the top when the window is scrolled).

* I think we can handle height resizing in CSS in this case (so I don't think we would need a resize event handler, but would have to see how it shakes out).

I think position fixed would be the easiest thing to debug and have by far the least CSS, HTML and JS. I am always reticent to use JS to calculate positions for items, but I think it's unavoidable in this context and I think the positional JS we need would be very lightweight.

# Thoughts

* We all like the idea of document previews in search.
* We are all agreed on the actual behaviour of what we are aiming for, as defined above.
* We should acknowledge we are describing is actually quite complex behaviour for a web app.
* If we have a solution that has clear limitations around it's implementation, that's okay and we shouldn't let that stop us from addressing the use case as captured above.
* It contends with UX problems most web sites don't go near and is in the gray area of where it does 'app' like things, but in a 'browser', which creates it's own challenges with regard to user expectations (such as 'where to scroll bars go' in certain views). There are not always clear answers to these.
* We should do user testing at the earliest opportunity.

**Comment by iaincollins** [comment]
@pudo if you have any comments about the format of arguments the # link or if any of the top level requirements are missing something that standards out just let me know :)

Obviously consistency for # path args makes sense, but don't know if we have a convention for them in mind already.

(( The CSS implementation bit is just for posterity / historical reference for anyone that wants to nerd out on the details - will be more succinct by next week. ))

**Comment by pudo** [comment]
As you say: it's a pretty rad thing to have a browser do, so I think the lesser of evils is a fair outcome. As long as we do it in the Screen component, we can refactor it much more easily than if we made the individual screens aware of the preview component and incorporated it into their layout.

I really like the proposed path schema, i.e. preview:type=document&preview:id=8277&preview:tab=tags etc. 

Looking at the code this weekend, I began to wonder if we want to declare the factsheets in the sidebars (e.g. the listing of all `DocumentMetadata`, source metadata (`CollectionInfo` 1 & 2), etc. etc.) to be `Viewers` in the `PdfViewer` sense. That would leave us with another root folder and help organize things in terms of file hierarchy depth as well as re-usability.

Maybe as a test run: shall we try to do the collections preview today? That should be the easiest one, I think.

**Comment by pudo** [comment]
Should we close this? There's obviously intra-doc and possibly some factoring out to do, but those could be separate tickets.

**Comment by iaincollins** [comment]
Yes I think so! No bugs in behaviour and the code isn't too terrible since recent improvements and I think all the things we have it are specific enough to be separate things.

**Comment by iaincollins** [comment]
tl;dr for future followers of this issue:

Using 'position: fixed' and just calculating for Navbar and Footer height and tracking scroll has been very performant and simpler than using (often quite buggy and complex) libraries that tried to wrangle 'sticky' into working properly. Sticky played horribly with flex and just did not work for us at all.
